### DISCUSSION TITLE
Unterstützung JK-BMS

### DISCUSSION THREAD

**Discussion by alexseuf** [shining-man/bsc_fw#6] [General]
Hallo,
gibt es eine Möglichkeit das JK-BMS zu unterstützten?
Gibt es schon funktionierende Platinen irgendwo zu kaufen?

**Comment by shining-man** [comment]
Theoretisch ist es möglich die Unterstützung für das JK-BMS zu integrieren. Hier wäre dann die Frage, ob über Bluetooth angebunden oder via RS485?
Je nachdem wie das Interesse an dem Projekt ist, kann ich mir die JK-BMS Integration einmal anschauen. Nur testen kann ich es nicht, da ich selber kein JK-BMS habe.

Fertig bestückte Platinen gibt es nicht zu kaufen. Ich habe noch 2-3 unbestückte Platinen, die ich abgeben kann, wenn ernsthaftes Interesse besteht.

**Comment by alexseuf** [comment]
Hallo Tobias,

erst mal danke für die schnelle Antwort:
mein mittelfristiger Plan ist es 2p16s System über 2 JK-BMS über einen
ESP32 zu einem Pylontech-Can-Protokoll zusammenzufassen, evtl. noch mit
einem Victron smart shunt und das ganze dann einem Deye 12kW WR am
BMS-eingang zur Verfügung zu stellen . Zusätzlich sollte das System noch
die Möglichkeit haben die Daten von einem Stromzähler (SDM630) abzugreifen
(und manipulieren zu können) bevor dieses dem WR zugeführt werden und auch
die Betriebsdaten für das Datenlogging vom Deye Wechselrichter zu sammel
und einer Datenbank mit Grafana Auswertung zur Verfügung zu stellen. (Davon
habe ich noch am wenigsten Ahnung.)

Die Daten von dem JK-BMs  sollen meiner Meinung nach über zwei je
potentialgetrennte Kabelverbindungen kommen (eine Bluetooth Verbindung
halte ich nicht für dauerhaft stabil. insbesonder nicht gut, wenn auf Basis
dieser Informationen Entscheidungen im Wechselrichter getroffen werden.
Im ersten Schritt würde es Sinn machen die RS232-Schnittstelle des JK-BMs
auszulesen. Hier sind meiner Meinung nach alle Daten für ein Pylontech 1.0
Protokoll enthalten.

wenn man es noch besser machen will (Pylontech 1.3 Protokoll)
http://forum.drbacke.de/viewtopic.php?p=69920&sid=cef2e5ea957521540279911f8b93d3a1#p69920
bräuchte man noch die Einzelwerte der Zellen, und Temperatursensoren, diese
sollte man über die LCD Schnittstelle der JK-BMs rausbekommen..
die Idee hate ich auch schon mal hier gepostet:
http://forum.drbacke.de/viewtopic.php?p=70090&sid=cef2e5ea957521540279911f8b93d3a1#p70090

Deine Hardware kommt mir von allen DIY Projekten die ich mir bisher
durchgeschaut habe, am sinnvollsten vor (Phönix-Steckverbinder)
potentialgetrennte RS485(2x)/RS232, keine gesteckten Module; alles gelötet,
Anschlüsse für DS18B20... einseitige SMD-LK wobei mir die Potentialtrennung
für die zweite RS232/Can Bus schnittstellen fehlen würde und ein 4 poliger
stecker für einen drehzahlgeregelten Lüfter mit Feedback  und ein
Schaltnetzteil so das die PCB direkt mit 12..48V versorgt werden kann.

Auch halte ich die Firmware für sehr gelungen insb. den Teil für die
Ansteuerung der Leistungsreduzierung des WR und das Webinterface

Wenn du mir zwei leer LK zur Verfügung stellen würdest, würde ich dort
gerne alle für die oben genannte Funktionserweiterung notwendigen Bauteile
bestücken und kann dir auch gern in China 10 neue leer LK bestellen..
Ich habe zwei JK-BMS, eines in der 1. Batterie(auch noch ein testaufbau)
und eines auf dem Labortisch, so das ein Test einfach möglich ist..  und
wenn es dich interessiert, kann ich auch gerne noch eines bestellen, das
ich dir leihe.

Dein Code sieht sehr sauber geschrieben aus, wobei ich mit dem Thema
Echtzeitbetriebssystem (Semaphore/tasks bisher noch nichts selber gemacht
habe.

Ist halt die Frage ob die idee die ich hier genannt habe auch in die
Richtung geht, die du gehen möchtest..
Denn ohne deine Hilfe werde ich mir sehr schwer tun, das Projekt auf Basis
deines Codes alleine so umzusetzen..
wobei das JK-BMS und der DEYE Wechslerichter meiner Meinung nach aktuell
der Platzhirsch in der DIY- szene sind..
Können auch gern mal telefonieren..

Mit freundlichen Grüßen



Am Fr., 28. Okt. 2022 um 19:41 Uhr schrieb shining-man <
***@***.***>:

**Reply by shining-man** [reply]
Zur Hardware:
Auf der LP sitzt ein DC/DC Wandler (U1) der hat einen Eingangsspannungsbereich von 4,5 bis 36 VDC. Dieser kann auch mit einem LCB78_03-0.5 bestücken werden, dann ist der Eingangsspannungsbereich von 4,5 bis 55 VDC. 

Es gibt in Summe drei serielle Schnittstellen die unterschiedlich bestückt werden können.
- U1: U4 (ADuM1201) oder U8 (SP3485)
- U2: U6 (ADuM1201)
- U3: U5 (ADuM1201) oder U9 (SP3485)
D.h. es können alle drei galvanisch getrennt betrieben werden. Es kann aber nicht gleichzeitig auch noch der RS485 Transceiver bestückt werden. D.h. entweder galvanisch getrennt oder RS485. Über die Lötjumper JP1 – JP6 können die RX/TX Signale vom ESP32 direkt heraus geführt werden.

CAN-Bus:
Den könnte man auch noch mit z.B. einem ADuM1201 galvanisch trennen. Das Problem ist momentan überhaupt noch einen ADuM1201 zu bekommen. 

Zur Lüftersteuerung:
Der Controller verfügt auch über einen I2C-Bus auf einem der Phönix Stecker. Diesen habe ich für Erweiterungen vorgesehen. Darüber wäre zum Beispiel eine Erweiterung für einen x-kanaligen PWM Regler denkbar.

Die Leiterplatte ist zweiseitig, da nicht alles für die DI’s auf der Top-Site untergebracht habe.

Zu Software:
Welche Schnittstellen stehen denn an dem JK-BMS zur Verfügung? Gibt das JK-BMS auch 3,3V/5V aus um die ADuM1201 sekundär mit Spannung zu versorgen?

Die Unterstützung für das JK-BMS an der seriellen Schnittstelle ist machbar und kann ich umsetzten. Ich habe ein Protokoll-Beschreibung, an der auch die Spannungswerte der Zellen enthalten sind.

Die Daten der zwei JK-BMS zusammenzufassen und via CAN an den DEYE zu senden, sehe ich auch nicht als Problem, solange der DEYE, das Protokoll der Pylontechs versteht.


Schreib mir einmal eine Mail an die shiningman@mail.de, damit ich deine E-Mail-Adresse habe.

**Reply by shining-man** [reply]
Noch einmal zur Spannungsversorgung:
Wichtig zu beachten ist, dass die Versorgungsspannung der Relais die Eingangsspannung ist. Da ich 5 V Relais verbaut habe, muss auch die Eingangsspannung 5V sein. Die Relais gibt es auch noch als 12 V Relais.
Alternativ könnte man die LP ändern, damit die Relais getrennt versorgt werden können.

**Reply by alexseuf** [reply]
den 5V auf 3V3 LDO würde ich erst mal so lassen und vorher halt ein (potentialtrennenden) 48V auf 5V DC/DC wandler anschließen.. bei einem LK redesing kann mann den dann auch integrieren (muss mann aber nicht)

**Comment by shining-man** [comment]
Link zum Communication Protocol vom JK-BMS:
http://www.jk-bms.com/Upload/2022-05-19/1621104621.pdf

**Comment by flashy02** [comment]
Hallo Tobias, hallo Alex,
Spannende Diskussion. Geht in die Richtung was ich mir aktuell auch vorstellen kann. Bin dabei meine laufende Victron Anlage mit Pylontech Akku um einen DIY mit JK BMS zu erweitern. Der erste Plan war, nur den Pylontech Akku am BMS dran zu lassen und den parallelen DIY nur mittels MQTT zu monitoren (dadurch bin ich beim googlen hier auf die Diskussion gestoßen), was aber bedeutet nur die Stromgrenzen des Pylontech nutzen zu können welche er mittels CAN an das VenusOS übermittelt. Schöner wäre es natürlich die Werte des Pylontech per Serial auszulesen um die Werte des JK zu ergänzen und dann die Summenwerte mittels CAN an das VenusOS zu senden. Gibt es denn die Platine zu kaufen?  Viele Grüße Michael

**Reply by shining-man** [reply]
Wenn sich die Pylontechs seriell auslesen lassen, dann lässt sich das Addieren der Daten realisieren. Hast du das Protokoll für die serielle Kommunikation mit den Pylontechs?

Zu den Leiterplatten: 
Ich habe keine mehr, die ich weitergeben kann. Ich kann dir die Gerberdaten geben, dann kannst du z.B. bei JLCPCB welche fertigen lassen. Das Kicad Projekt der PCB ist auch auf github. 
Aktuell mach ich ein kleines Redesign, was aber sicher noch ein paar Wochen dauern wird. Es wird sich mit dem Redesign aber nichts Grundlegendes ändern. Es betrifft eher die Potenzialtrennung der Schnittstellen.
Alexander hat einen Reichelt-Warenkorb (https://github.com/shining-man/bsc_fw/discussions/7#discussion-4524933) mit allen Bauteilen (außer den gängigen SMD Widerständen/Kondensatoren) erstellt.

Wenn mehr Interesse an den Leiterplatten vorhanden ist könnte man das Redesign evtl. auch bestücken lassen. Um die Fertigung und die Verteilung untereinander (Mindestabnahmemenge sind, glaube ich, 5 Stück) muss sich aber jemand kümmern.

**Reply by flashy02** [reply]
Das klingt ja super, aber dann warte ich noch auf das Redesign. Ist aktuell nicht zeitkritisch. Bis mein Akku läuft dauert auch noch ein paar Wochen, noch warte ich auf die Zellen. Würde dann im ersten Schritt versuchen dein Projekt zu kompilieren und auf nen Esp drauf flashen. Wenn ich nem anderen Repo trauen kann, kann ich den TTL direkt ohne Galvanische Trennung an den ESP anschließen und das JK BMS auslesen. Im ersten Step genügt es mir die Daten alle in mein Openhab per MQTT zu integrieren und zu überwachen.

**Reply by flashy02** [reply]
P.S. bezüglich Schnittstellenbeschreibung muss ich in meinem Fundus schauen, glaube da hab ich was. Aktuell hängt bei mir nen ESP8266 am RS232 und liest alle Werte aus und sendet sie an mein Openhab. Über den CAN bzw. Das VenusOS kommen nur die rudimentären Daten wie Spannung, Strom, Temp und Spannungsgrenzen. Die einzelnen Zellspannungen meiner 8 Pylontech Module (8×15=120) leider nicht.

**Comment by alexseuf** [comment]
Hallo flashy,
Wenn du erst Mal nur eine Verbindung vom ESP zu einem JK BMS hast und ein
potentiallgetrentes Netzteil verwendest ( keinen Buck Konverter aus der
Batterie!) dann kannst du das so machen..
Aber sobald du weitere galvanische Verbindung nach außen machst, wie CAN
Bus dann ist das keine gute Idee..
Du brauchst auch nicht auf das Redesign der leiterkarte zu warten.. du
kannst auch für erste Tests einfach eine esp32 node mcu bei Amazon
bestellen..  https://amzn.eu/d/dFsBpOJ
Und ein fertig compiliertes binary drauf spielen...   Bei der Verdrahtung
der drei Leitungen zum JK BMS kann ich dir dann Tipps geben.

flashy02 ***@***.***> schrieb am Sa., 5. Nov. 2022, 07:49:

**Reply by flashy02** [reply]
Moin Alex,
Wieso sollte ich keinen Buck Converter nutzen um den Esp zu versorgen? Was bietet denn das JK am vbat Anschluss des Seriellen Anschluss für eine Spannung an? Wirklich vbat oder eher 5v? Wenn ich später über CAN wirklich eine Pylontech Simulation laufen lassen werde für die Kombination der Stromgrenzen sieht es natürlich anders aus. Aktuell versorgt auch die cerbo über den USB meinen ESP mit welchem ich den Pylontech auslese.
MfG Michael

**Comment by alexseuf** [comment]
Ein normaler Buck tiefstsetzsteller bietet keine potentialtrennung. Dein
jk-BMS Trennt im Fehlerfall den batterie minus pfad. Bei den pylontech
werden den Fehlerfall soviel ich weiß der Batterie plus Weg getrennt. Über
die nicht potentialfreien ground oder v-bat Verbindungen schaffst du
parallelwege die nicht getrennt werden über die dann die Ströme fließen für
die diese Verbindung nie gedacht waren.
Schaut dazu mal im dr Backe Forum zum Thema defekte RS485 Umsetzer für
jkbms.
Ansonsten hilft eine potentialtrennung auch immer Störungen zu vermeiden
also auch Ac Störungen und nicht nur DC ground loops zu verhindern. Beim
System in der kostenordnung wie das batteriesystem würde ich nicht an den
10 € für ein Potenzial getrenntes Netzteil sparen.. wenn du wirklich alles
bis ins letzte, inklusive aller möglichen Fehlerfälle betrachtet hast, dann
kann man das sicherlich machen. Aber vermutlich ist es da einfacher die 10
€ mehr zu investieren..
Schau dir dazu auch noch mal meine Beiträge im Forum zum Thema Absicherung
der Netzteile an.. wenn du sowas direkt an die Batterie hängst und das
Netzteil kaputt geht dann kannst das dann eine ganz schön Schaden
anrichten..

Beim JK BMS gilt: wo v-bat drauf steht ist auch vbat drinnen. Aber messe es
besser zur Sicherheit noch mal genau an deinem Modell nach.. (
multimeter).. kannst dann auch noch mal den gnd Bezug der verschiedenen
Anschlüsse in Bezug auf B- bzw P- messen. Für den Fall dass das BMS den
hauptstrompfad geschlossen oder geöffnet hat.

Will dich mit den Aussagen nicht ärgern ist einfach nur meine Meinung.. und
auch ein bisschen Erfahrung der letzten Jahrzehnte in meinem Beruf..


flashy02 ***@***.***> schrieb am Sa., 5. Nov. 2022, 09:25:

**Reply by flashy02** [reply]
Hallo Alex,
Passt. Es geht also um das Risiko im Fehlerfall wenn das BMS "Aufmacht". Da ich davon ausgehe, dass sich das BMS dadurch nicht selbst trennt, sonst würde es ja die Kontrolle verlieren, werden weiterhin sowohl Ground als auch VBat an den Ausgängen liegen. D.h.wenn ich die Versorgung über das was das BMS zur Verfügung stellt realisiere und keine Verbindung zu einem anderen Controller herstelle, ist das Risiko überschaubar. Wieso bist du der Meinung du ärgerst mich mit deinen Aussagen? Deine Einwände sind ja berechtigt. Die gleichen Einwände diskutieren wir aktuell auf der Arbeit bezüglich Entwicklung eines GNG Ladegerätes.

**Comment by alexseuf** [comment]
Beim JK BMS habe ich allerdings keine Angabe gefunden welcher maximalstrom
über v-bat gezogen werden kann und ob das irgendwie abgesichert ist außer
über die leiterbahn der leiterkarte.. bei chinesischen Produkten bin ich da
immer etwas vorsichtiger..
Im Forum hatten viele empfindlich auf mein Argumentation reagiert da sie
anscheinend alle fleißig buck wandler verbaut hatten.. und sich das ganze
nun schön reden wollen.. und die Ausfälle der RS485 Schnittstelle kommen
sicherlich von der mangelnden Qualität von JK... Und nicht von der eigenen
Auslegung..

Was ist denn eine gng-ladegerät?
Grüße Alexander

flashy02 ***@***.***> schrieb am Sa., 5. Nov. 2022, 10:07:

**Reply by flashy02** [reply]
Ein Galvanisch nicht getrenntes Ladegerät, also 230V AC auf >400V DC. Will aber hier jetzt nicht die Diskussion von Tobias sprengen.

**Comment by TwisterSB** [comment]
Hallo zusammen.
Ich schreibe dann mal hier, nachdem ich mir jetzt auch mal einen Account gemacht habe.
Ich bin im Forum (drbacke) auf dieses geniale Projekt gestossen. Respekt was du hier auf die Beine gestellt hast.
Kurz zu meiner vorgesehenen Anwendung. Deye 12kw WR, JBD BMS, Neey 4A Balancer. Das sollte ja jetzt schon alles implementiert sein.
Ich lasse öfter mal Platinen bei JLC machen, mal unbestückt, mal bestückt. Da hier ja viele Bauteile drauf sind, würde ich natürlich zu bestückt tendieren. Am Wochenende hab ich mir mal die Zeit genommen, und die BOM mit Teilen die bei JLC verfügbar sind gefüttert. Für eine beidseitige Bestückung würde ich auf einen Preis von rund 400$ für 10 Stück kommen. Da wäre dann wirklich alles (auch ESP) bis auf den DC DC dabei, den hatte ich auf die schnelle nicht gefunden.
Jetzt habe ich hier gelesen, das du noch ein Redesign planst. Dann würde ich natürliche nochmal abwarten, und werde dein Projekt mal weiter mit Spannung  verfolgen.

**Reply by flashy02** [reply]
Hallo TwisterSB, falls du bei JLC was in Auftrag geben würdest und du noch Abnehmer benötigst. Hätte Interesse an zwei bestückten PCBs.

**Reply by shining-man** [reply]
JBD-BMS und NEEY wird unterstützt und funktioniert. Die beiden habe ich selber in Verwendung.
Ich habe zwar keinen DEYE Wechselrichter, da meiner aber auch das Pylontech-Protokoll versteht, gehe ich davon aus, dass es an einem DEYE genauso funktioniert. Wenn nicht, dann bekommen wir das schon zum Laufen. 
Genau, ich mache noch einmal ein Redesign, was vor allem die galvanische Trennung der Schnittstellen betrifft. Ich möchte dies auch in den nächsten Wochen fertig bekommen.
Gerne können auch noch Ideen oder Anregungen kommen für das Redesign.

Nutzt du evtl. mehr als einem NEEY Balancer?

**Reply by kmn13** [reply]
Hi zusammen ;-)
@shining-man zuerst einmal großartiges Projekt!
Bin gestern bei der Suche über Möglichkeiten zum NEEY auslesen auf dieses geniale Projekt gestoßen. Es deckt sich bis auf das BMS und die Anzahl der NEEYs mit meinem Plan / Idee.

**Comment by TwisterSB** [comment]
So als spontane Idee für ein Redesign würde mir als erstes nur ein paar Beschriftungen an den Anschlüssen einfallen. Das wird die Platine zwar etwas grösser machen, aber bei den Preisen der Platinenfertigung in China fällt das nicht ins Gewicht. Das wäre aber nur meiner Faulheit geschuldet, ich bin immer zu faul mir bei meinen Projekten den Schaltplan beim anklemmen dazu zu nehmen, daher beschrifte ich jeden Kontakt wo was externes angeschlossen wird :-)
Aufgefallen war mir das der Footprint für die BAT54W an D20,D22,D21,D23 nicht ganz passt, aber das habe ich dann gesehen, das du das ja schon im Schaltplan vermerkt hast.

Ich nutze nur einen NEEY Balancer. Aber wenn es dir darum geht, zu testen ob die Kommunikation mit mehreren funktioniert, ich habe Zugriff zu zwei weiteren NEEY Balancer von Kollegen.

Selber brauche ich nur 3 Platinen, eine für mich, und 2 für Kollegen. Aber ab 10 ist der Preis interessanter. Wenn gewünscht könnte ich dann wenn es soweit ist eine kleine Sammelbestellung machen. Selbstverständlich zum Selbstkostenpreis.

**Reply by shining-man** [reply]
Die Höhe der Platine ist 72 mm für das Phoenix Profil UM 72. Dies möchte ich eigentlich auch so belassen, da es sich damit einfach und einigermaßen sauber auf Hutschiene montieren lässt.

Bei dem NEEY Balancer ginge es mir genau darum, ob die Kommunikation mit mehreren funktioniert. Wenn du das mit drei testen könntest, wäre das super. Um dies zu testen, genügt auch ein einfaches ESP32 Board.

Wenn sich genug Interessenten finden, ist eine Sammelbestellung eine gute Sache. Vor allem entfällt die meiste Lötarbeit:)

**Reply by TwisterSB** [reply]
Die Kommunikation mit mehreren NEEY testen kann ich dann übernehmen, kann von meinen 2 Kollegen den NEEY dann mal bekommen, wenn die Teile da sind.
Auf die Akkus und BMS warten wir aber auch noch, Test wäre dann also erstmal ohne echte Daten der Zellen.............

Das mit dem Phoenix Profil macht natürlich absolut Sinn, ich verwende das auch gerne, allerdings habe ich die 100mm Variante. Vielleicht könnte man einfach 2 Versionen machen, 72mm ohne Beschriftung, und 100mm mit Beschriftung, die 14mm pro Seite sollten da schön passen, ohne das man dann grundsätzliche Änderungen im Layout macht.

**Reply by shining-man** [reply]
Wenn du beim Testen Fragen hast oder es zu Problemen kommt, kannst du mich auch per Mail erreichen.

**Reply by kmn13** [reply]
@TwisterSB
Also falls es wirklich die Möglichkeit gibt, sich an einem Leiterplatten Auftrag zu beteiligen würde ich gerne 2-3 Stück mitabnehmen (wir sind zu 2. und bauen 2 idente Systeme, 1 Reserve schadet ja nie ;-))

Das Redesign klingt spannend - speziell die Version mit dem entkoppelten CAN Bus finde ich gut.

**Comment by flashy02** [comment]
Hallo Tobias,
Wenn ich das richtig verstehe, wird das JK BMS aktuell über die serielle Schnittstelle ausgelesen. Ist es auch geplant dies in Zukunft über die BT Schnittstelle einzubinden? Z.B. wie mein Arbeitskollege das hier im Forum umgesetzt hat? Hier kann man mehr Informationen auslesen wie rein über serial, z.B. den Balancer Strom.
https://forum.drbacke.de/viewtopic.php?t=982&sid=0ce19c8981bb5366c11b3b822cdc3f9c

**Reply by shining-man** [reply]
Erst einmal ist es nicht geplant das ganze über Bluetooth auszulesen. Eine Verbindung über RS485 ist meiner Meinung nach viel stabiler. Vor allem wenn die Daten dafür genutzt werden, um diese über CAN an den Wechselrichter weiterzuleiten, ist eine stabile Verbindung wichtig.
Wenn aber Bedarf an der Bluetoothanbindung des JK-BMS besteht, kann z.B. einfach ein Issue aufgemacht werden, dann kann ich dies bei Gelegenheit & Zeit implementieren. Voraussetzung wäre in dem Fall, dass ich auch ein Communication-Protocol habe. Da ich selber kein JK habe, müsste auch jemand bereits sein die Entwicklungsstände zu testen.

**Reply by flashy02** [reply]
Bezüglich Verbindungssicherheit gebe ich dir recht. Nichts geht über eine Kabelverbindung. Für die kritischen Umfänge auf jeden Fall. Wollte mir dann auch noch anschauen wie das JK reagiert, wenn man parallel sowohl über BT als auch Serial das BMS abfragt, ggf. könnte man dies dann auch kombinieren. Ist aber erst einmal Zukunftsmusik. Werde dann ggf. Ein Issue eröffnen, wenn der Bedarf dann noch besteht. Bezüglich Testen der Entwicklungsstände wäre dann nicht das Thema, bzw. würde ich übernehmen. Aber erst benötige ich eine funktionstüchtige Batterie und die bestückte/zusammengelötete Platine. Das bedarf noch ein paar Wochen.

**Comment by drakesoft** [comment]
Hallo zusammen, 
erst mal vorweg. Super Projekt. Ich hätte mal angefangen die Firmware auch für die JK BMS JK-B1A24S/JK-B2A24S anzupassen, die haben ein anderes Protokoll.
Aktuell scheitere ich leider an dem was der UART empfängt. Senden ist ok das BMS antwortet darauf. Über Rs485 und am RX pin ist die antwort korrekt. Aber im UART empfangs buffer habe ich nur grütze stehen.

Das kommt an:
Timeout: Serial=0, u8_lRecvDataLen=74, u8_lRecvBytesCnt=0
0d 80 c0 20 ff 3f 8c 3f fc 3f ff ff 3f b3 34 61 09 80 30 ce fb 3f 03 00 00 00 23 00 06 00 7a 42 0d 80 e0 2p ff 3f 8f 04 40 3f 8c 3f fc 3f a8 de fb 3f ab ab 00 00 ff ff 3f 00 ff ff ff ff 4e 8b 0d 80 00 21 ff gf b4 45 fc 3f

Und ankommen sollte:
EB 90 01 FF 1E D3 0F 69 14 13 ....

Hatte von euch jemand schon mal sowas?

[JK-B1A24S-RS485-protocol-english.pdf](https://github.com/shining-man/bsc_fw/files/10197919/JK-B1A24S-RS485-protocol-english.pdf)

**Reply by shining-man** [reply]
Schaut nach falscher Baudrate aus. Laut deinem Protocol arbeitet dein BMS mit 9600 Baud, das JK-B2A20S20P mit 115200 Baud.

**Reply by drakesoft** [reply]
ne Baudrate is schon auf 9600 umgestellt ... senden würde mit der falschen ja auch nicht funktionieren.

**Reply by shining-man** [reply]
Auf den ersten Blick ähnelt das Protokoll mehr dem JBD-BMS als dem JK-BMS. Ich würde als Basis das JBD nehmen.

**Reply by drakesoft** [reply]
uff kA kann schon sein ... wenn ist es jetzt zu spät. Kannst mal drüber schaun ob dir was ins auge springt und stecken bleibt. :-D

[bsc_fw-main_JK-DZ11-B2A24S.zip](https://github.com/shining-man/bsc_fw/files/10198015/bsc_fw-main_JK-DZ11-B2A24S.zip)

**Reply by shining-man** [reply]
Jetzt verstehe ich auch warum du soviel "Müll" zurückbekommst.
Das Problem ist, dass in der V0.1.5, die du als Basis genommen hast, die Debugausgaben über einen SW-Serial mit 115200 Baud erfolgen. Du mußt die Baudrate für die Debugausgabe auf 19200 reduzieren, sonnst kommt da nichts vernünftiges. Meine Empfehlung wäre, auf die aktuelle Version aus dem main branch zu gehen (V0.2.5). Ist zwar noch nicht released, aber läuft bei meinen Tests bisher stabil. Zwischen 0.1.5 und 0.2.5 hat sich, gerade im unterbau, einiges geändert. Deine neuen BMS-Files solltest du aber größtenteils verwenden können. Debugausgaben erfolgen jetzt mit ESP_LOGD(...). Du kannst in der defines.h dann festlegen ob er für die Debugausgaben einen HW-Serial nehmen soll. Dann ist nur zu beachten, dass der Serial 2 des BSC zu einem SW-Serial wird, was aber zum Entwickeln kein Problem ist, da darf dann halt nicht dein BMS angeschlossen sein. Zusätzlich kannst du in der defines.h sagen ob er in ein Log-File loggen soll. Würde ich bei hohem TX evtl. deaktivieren.
```
defines.h
//#define DEBUG_ON_FS
#define DEBUG_ON_HW_SERIAL
#define LOG_TO_SERIAL
```

**Reply by drakesoft** [reply]
Ok danke ... dann verpflanz ich das mal in die aktuelle Dev version.

**Reply by drakesoft** [reply]
Die RS485 mit dem JK BMS funktioniert jetzt. Dafür mag die CAN Schnittstelle nicht. Laut log hat der init funktioniert und die funktion sendCanMsg hab ich mal mit ner debugmeldung versehen, die wird auch aufgerufen... jedoch kommt auf RXD und TXD nichts raus außer beim Start mal kurz auf low und dann bleibt RXD und TXD auf high.
Ist das erste mal das ich mit dem ESP32 was mache... hast du nen tipp was da schief läuft?

![Screenshot_20221215_204130](https://user-images.githubusercontent.com/8250445/207952450-ecf72c7a-68c9-441d-b504-2e1724af5a2a.png)

**Reply by shining-man** [reply]
Wie ist denn der Stromlaufplan vom Aufbau? 
Hast du auch einen CAN Transceiver an den entsprechenden GPIOs angeschlossen oder hängen die in der "Luft"?

**Reply by drakesoft** [reply]
Transceiver hängt an GPIO4/5 und an CANH/L hängt der WR.

**Reply by shining-man** [reply]
Was mir noch noch eingefallen ist. Als Basis verwende ich folgende CAN Lib: https://github.com/sandeepmistry/arduino-CAN
Da diese mit meiner ESP Revison aufgrund eines Fehlers in der CAN-Lib nicht funktioniert hat, musste ich diese anpassen. Probiere es einmal mir der original CAN-Lib Version von github.

**Reply by drakesoft** [reply]
Danke für deine schnelle Antwort... am ende wars jetzt der CAN transceiver. Wie auch immer jetzt funktioniert die Kommunikation mit dem WR und dem BMS. Jedoch verhält ich der WR merkwürdig ... schaltet immer zwischen Laden und Endladen hin und her ohne Grund. Eigentlich sollte er Laden da die max. Ladespannung nicht erreicht ist. Ich hab einen DEYE Hybrid. Kommt dir das bekannt vor? Sind noch daten falsch? 
Was mein JKBMS nicht kann ist Strom messen ... also ist getBmsTotalCurrent() = 0 (CAN ID 0x356) und den SOC muss ich anhand der Spannung ausrechnen / abschätzen.

**Reply by shining-man** [reply]
Aus dem Stehrgreif fällt mir nichts ein. Du kannst mir deinen Code schicken, dann schaue ich mir ihn einmal an, ob ich etwas auffälliges sehe. 
Du kannst dir auch einmal die CAN-Messages als Log ausgeben lassen und schauen ob hier die Werte stabiel sind.

**Reply by drakesoft** [reply]
Hi, 
ich weiß nicht vorhin hab ich mir eingebildet das es ging aber jetzt Endlädt er nicht ... kA warum. 
Was ich noch nicht ganz verstanden habe ist wie man die Mindestzellspannung einstellt bis zu der Endladen werden darf, geht das über den Alarm? 
Zum Code: Was ich grob gemacht habe....
1. einen Include gefix, Linux hat case sensitives filesystem und compiliert sonst nicht
2. in sendCanMsg_355() sollte glaube ich die einstellung ID_PARAM_INVERTER_SOC_BELOW_ZELLSPANNUNG_EN vom Webinterface beachtet werden ... ich hab die entsprechende if hinzugefügt
3. die Unterstützung für das oben genannte JKBMS hinzugefügt. 
4. Die Limits für eine handvoll werte geändert da ich LTO Zellen verwende und die eine niedrigere Spannung haben.

Möchtest du die Änderungen übernehmen wenns fertig ist und funktioniert?
[bsc_fw.zip](https://github.com/shining-man/bsc_fw/files/10254405/bsc_fw.zip)

**Reply by drakesoft** [reply]
Vergessen aktuelle Debugmeldungen:
D (3765290) JK_BMS_V13: Vstack=5650
D (3765290) JK_BMS_V13: Vavgcell=2368
D (3765290) JK_BMS_V13: BMSerror=0
D (3765290) JK_BMS_V13: Vcdiff=2368
D (3765292) JK_BMS_V13: VbalC=1023
D (3765295) JK_BMS_V13: VbalA=1
D (3765298) JK_BMS_V13: CellCount=24
D (3765301) JK_BMS_V13: TotalV=56835, MaxV=2377, MinV=2360
D (3765306) JK_BMS_V13: SOC=75, SumV=56835
D (3765310) JK_BMS_V13: BMStemp=7
D (3765412) CAN: Soll Ladestrom: 240, 240, 240, 240
D (3765412) CAN: 0x351 ChargeCurrent=2400 DischargeCurrent=2400 ChargeVoltageLimit=610
D (3765419) CAN: 0x355 soc=75
D (3765424) CAN: 0x356 current=0 temperature=70 voltage=5683
D (3766313) JK_BMS_V13: Serial 0 send

**Reply by drakesoft** [reply]
Jetzt gehts ... war ne einstellung vom WR ... die timetable funktion hat mich ausgetrickst.

**Reply by shining-man** [reply]
Wenn alles läuft, würde ich den Code für das BMS übernehmen.
Hast du eigentlich als Wechselrichter DEYE oder Victron ausgewählt?
Das Entladen geht jetzt. Das Laden hat aber hatte aber auch Probleme gemacht, oder? Meinst du das die Einstellung sich auch auf das Laden auswirkt?

**Reply by shining-man** [reply]
> Was ich noch nicht ganz verstanden habe ist wie man die Mindestzellspannung einstellt bis zu der Endladen werden darf, geht das über den Alarm?

Würde über einen Alarm gehen. Das BMS auf die Min. Zellspannung überwachen und mit dem Alarm die Entladeleistung auf 0 stellen.
Wenn hier eine direkte Funktion gewünscht ist, die noch nicht vorhanden ist, dann kann ich das einfügen. In dem Fall am besten einfach ein Issue aufmachen.

**Reply by drakesoft** [reply]
Ich denke das waren zwei probleme ... das  Laden / Endladen problem war glaube ich ein Skalierungsfehler von der Gesamtspannung. Das nicht mehr Endladen Problem war die Einstellung am WR. In der Timetable war auf 80% SOC gestellt und genau das hat er gemacht auf 80% Endladen und dann war schluss ... deswegen dachte ich auch erst das es jetzt gehts aber da war ich noch über 80% :-D. 
Wenn du den Code übernehmen willst dann sollten wir uns gedanken machen wie wir das mit der SOC Berechung einstellbar machen. Z.b. wenn man in "Wechselrichter & Laderegelung" eine Einstellung/Funktion macht "Overwrite SOC"  mit min. und max. Spannung als Einstellung?
